docs/what-is-hls.md

```diff
@@ -0,0 +1,58 @@
+# What is haskell-language-server?
+
+The `haskell-language-server` (HLS) project is an implementation of a server (a "language server") for the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) (LSP).
+A language server talks to a client (typically an editor), which can ask the server to perform various operations, such as reporting errors or providing code completions.
+The advantage of this system is that clients and servers can interoperate more easily so long as they all speak the LSP protocol.
+In the case of HLS, that means that it can be used with many different editors, since editor support for the LSP protocol is now widespread.
+
+## Language Server Protocol basics
+
+### Servers and clients
+
+HLS is responsible for actually understanding your project and answering the questions that the client asks of it, such as: what completion items could go here? are there any errors in the project? and so on.
+HLS provides [many](./features.md) (but not all) of the features that the LSP protocol supports.
+
+But HLS only provides the server part of the setup.
+In order to actually use it you also need a client (editor).
+The client is responsible for managing your interaction with the server: launching it, dispatching commands to it, and displaying or implementing responses.
+Some clients will even install the server binaries for you!
+
+Common clients include:
+- VSCode (the reference implementation for a LSP client)
+- Emacs, with the `lsp-mode`+`lsp-haskell` or `eglot` packages
+- Vim/neovim, with the `coc.vim` or `LanguageClient-neovim` packages
+- Kate
+- ... and more every day!
+
+### LSP terminology
+
+Here are a few pieces of jargon that you may come across in the HLS docs or when discussing problems:
+
+- *Code action*: A code action is a specific action triggered by a user on a particular region of code. Examples might include "add a type signature to this function".
+- *Code lens*: A pre-rendered edit or action shown in the body of the document itself, usually triggered with a click. Examples might include "the type signature for a function, which is actually inserted on click".
+- *Completion item*: An item that can be inserted into the text, including its metadata.
+- *Diagnostic*: Any information about the project that is shown in the editor, including errors, warnings, and hints from tools such as hlint.
+- *Semantic highlighting*: Special syntax highlighting performed by the server.
+
+## The structure of HLS
```

Comment:
Oooh, is that a new section? Looks promising! Not sure whether you've seen it, but for GSoC 2020 bubba created a nice visualization: https://mpickering.github.io/ide/posts/2020-05-08-state-of-haskell-ide.html

Maybe we can create something similar and link it here?

Reply:
I think it's probably misnamed :joy: It's more "here's some stuff we talk about when we talk about HLS that's useful to know about", like plugins! and hie-bios! It's not _really_ an explanation of the structure...

Reply:
But it could be! I really loved https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/architecture.md which explains what each component does. However, what I have just linked is way more technical than our current section. But maybe some future work? Nevertheless, just wanted to say, I like the new section :D

Reply:
Future work :)